### PR TITLE
[FIX] Force the parts of PolyData to be 2D

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes none

Changes proposed in this pull request:

- make it explicit that left and right for PolyData must be 2D numpy array
  - We used to have SurfaceImage with shape (20484,): this is not allowed anymore and it will be SurfaceImage (1, 20484)
- add check to make that surface images passed for plotting have a single timepoint (shape = (1, x))
- adapt how 1D array are loaded / saved

### CODE REVIEW

File: nilearn/surface/surface.py
Comment (Remi-Gau) [inline nilearn/surface/surface.py:1252]: Something that could be nicer for users, is that if a 1D numpy array then it is converted to a 2D array.

@man-shu @bthirion WDYT?

Comment (man-shu) [reply nilearn/surface/surface.py:1252]: I see, so this would ensure that the data arrays are 2D, even if there is only one sample/timepoint.

1. Could cause some test failures on #4770. So we should merge this one first.
2. Would the saving/writing be unaffected? For example, would the .gii files saved by nilearn still work with other software?

Comment (man-shu) [reply nilearn/surface/surface.py:1252]: Otherwise I am ok with the changes.

Comment (Remi-Gau) [reply nilearn/surface/surface.py:1252]: > 1. Could cause some test failures on

Yeah I was having a look at what you were doing and thought it may cause issues.

Comment (Remi-Gau) [reply nilearn/surface/surface.py:1252]: > I see, so this would ensure that the data arrays are 2D, even if there is only one sample/timepoint.

Yes but I think it may be easier to use in general if we can still instantiate a Surface image with 1 dim numpy array for one hemisphere but we just force them to be 2 dim.

Comment (Remi-Gau) [reply nilearn/surface/surface.py:1252]: Will wait for the test to all pass before I refactor things.

File: examples/01_plotting/plot_surf_stat_map.py
Comment (Remi-Gau) [inline examples/01_plotting/plot_surf_stat_map.py:37]:
```suggestion
```

File: nilearn/surface/surface.py
Comment (Remi-Gau) [inline nilearn/surface/surface.py:1774]:
```suggestion
```

**Review by man-shu — APPROVED** [review]
I am for the change simply because coming from volumetric images, it took a while to wrap my head around why the data was in 1D format.

However, I can't help but notice that it introduces some code that might not initially be obvious to the users. For example, adding a `[0]` to fetch the data even when there's only one sample or adding an extra pair of square brackets.

It would also be inconsistent with NIFTI images where we have 3 dimensions when there's only one sample and the 4th dim only comes into the picture when time gets involved.

So I would be interested in hearing what other people think about this.

File: doc/visual_testing/reporter_visual_inspection_suite.py
Comment (Remi-Gau) [inline doc/visual_testing/reporter_visual_inspection_suite.py:469]: unrelated to the PR but realized this was already called above

File: examples/08_experimental/README.rst
Comment (Remi-Gau) [inline examples/08_experimental/README.rst:1]: forgot to remove it before

File: nilearn/datasets/struct.py
Comment (Remi-Gau) [inline nilearn/datasets/struct.py:1151]: `load_surf_data` will be called in `PolyData.__init__`

File: nilearn/glm/first_level/first_level.py
Comment (Remi-Gau) [inline nilearn/glm/first_level/first_level.py:1014]:
```suggestion
                        run_img.data.parts[part].shape[1], dtype=bool
```

File: nilearn/maskers/_utils.py
Comment (Remi-Gau) [inline nilearn/maskers/_utils.py:71]:
```suggestion
        data[part] = np.mean(value, axis=0).astype(float)
```

File: nilearn/surface/surface.py
Comment (Remi-Gau) [inline nilearn/surface/surface.py:1284]: The issue is that if a user set some parts manually after init and does not run _check_parts they may end up having invalid parts.

We already had this problem before, but this makes it more clear.

File: nilearn/surface/surface.py
Comment (Remi-Gau) [inline nilearn/surface/surface.py:1271]: Not too happy with this way of 'tracking' what the object was initialized with just to ensure that 1D array are saved properly to a gifti file, but at least it works...

Well unless, once again, some of the parts are manually set with a 1D array after init.

**Comment by Remi-Gau** [comment]
@bthirion @jeromedockes @htwangtw @emdupre 
any opinions on this?

File: nilearn/surface/surface.py
Comment (bthirion) [reply nilearn/surface/surface.py:1271]: `self._squeeze_on_save` should be a Bool: there should not be a distinction between left and right hemispheres.
And I think it should be configurable at init.

**Review by bthirion — APPROVED** [review]
LGTM, thx.

**Comment by Remi-Gau** [comment]
Will wait for CI to finish and merge

**Comment by jeromedockes** [comment]
sorry I took a few days to respond but what's the rationale behind this? when the dimension of the image is one value per vertex -- for example a mask, a non-probabilistic atlas, a statistical image or anything that isn't a time series -- what is the benefit of adding an extra dimension? that would be like making all volume images 4d, it hides the actual dimension of the data and means we have to check that the last length is 1 and squeeze it whenever we have a 1d image (similar comment to that of @man-shu [here](https://github.com/nilearn/nilearn/pull/4771#pullrequestreview-2450964339) )

**Comment by bthirion** [comment]
"Always 4D" makes things simpler in the sense that you don't need to figure out, but indeed we could also strive to have the best-suited data dimension.

**Comment by Remi-Gau** [comment]
An example of the weird things that would happen before this PR.

Ran the following after `git checkout ad79dadf0`

```python
from nilearn.datasets import load_fsaverage
from nilearn.surface import SurfaceImage
import numpy as np

mesh = load_fsaverage()["pial"]

# one time point image: impmlicit
data = {
    "left": np.ones((mesh.parts["left"].n_vertices)),
    "right": np.ones((mesh.parts["right"].n_vertices)),
}
img = SurfaceImage(mesh=mesh, data=data)

print(f"{img=}")
print(f"{img.shape[0]=}")

# one time point image: explicit
data = {
    "left": np.ones((1, mesh.parts["left"].n_vertices)),
    "right": np.ones((1, mesh.parts["right"].n_vertices)),
}
img = SurfaceImage(mesh=mesh, data=data)

print(f"{img=}")
print(f"{img.shape[0]=}")

# 2 time points image
data = {
    "left": np.ones((2, mesh.parts["left"].n_vertices)),
    "right": np.ones((2, mesh.parts["right"].n_vertices)),
}
img = SurfaceImage(mesh=mesh, data=data)

print(f"{img=}")
print(f"{img.shape[0]=}")
```

Gives me this:

```
img=<SurfaceImage (20484,)>
img.shape[0]=20484
img=<SurfaceImage (1, 20484)>
img.shape[0]=1
img=<SurfaceImage (2, 20484)>
img.shape[0]=2
```

> it hides the actual dimension of the data
To me I would say that this actually makes it explicit.

The analogy to volume does not quite work as the 'time' dimension comes last for volume data (x, y, z, **t**) but first for surfaces (**t**, v).

**Comment by jeromedockes** [comment]
not after #4591 which makes t the last dimension, so the first dimension would always be n vertices.

as a user I expect the dimensions of image to match those of the data I passed or that I have in my Gifti file. I don't really see why they would change or why I am asked if they should be squeezed on save: just save the data and shape of what I passed.

so IMO as the maskers (AFAICT) were working OK in 1, 2 or more dimensions, forcing it to be 2 is a regression. more than 2 may not be useful but 1D certainly is (an atlas is not a time series with a single time point: it does not have a time dimension).

in any case it's not a big deal. pr #4591 has been updated to match the "always 2D" choice and should be ok to merge now

**Comment by Remi-Gau** [comment]
> not after #4591 which makes t the last dimension, so the first dimension would always be n vertices.

yes true

> why I am asked if they should be squeezed on save: just save the data and shape of what I passed.

Agree that the squeeze on save is clunky.

> an atlas is not a time series with a single time point
When working with volume I would still stay that my image image contains a single volume

FYI I admit my matlab upbringing may be tainting my thinking on this:

```matlab
% create 3 array
>> a = ones(3,3,3);
% asking for the 4th dimension of that 3D array
>> size(a, 4)
ans =
     1
>> 
```

**Comment by Remi-Gau** [comment]
I'd be OK to maybe revert force to 2D after #4591 is in.

**Comment by bthirion** [comment]
I prefer to leave users control rather than coercing to 2D...

**Comment by Remi-Gau** [comment]
OK I will revert after #4591 is merged then

**Pull request merged**
